mlir/lib/Transforms/Utils/DialectConversion.cpp

```diff
@@ -1328,15 +1328,19 @@ Block *ConversionPatternRewriterImpl::applySignatureConversion(
     mapping.map(origArg, argMat);
     appendRewrite<ReplaceBlockArgRewrite>(block, origArg);
```

Comment:
Another thing that I noticed while debugging around here, the line 1329 are adding to the map and adding a pending block argument rewrite, and so is line 1349 below. Also the `builtin.unresolve_conversion_cast` generated in line 1325 is overriden by the one generated in line 1345 in some cases (with the mapping changed as well). I dont think that was intended. Might just be a harmless issue, but might be hiding a bug there.

Reply:
There are two mapping steps: `origArg -> argMat -> targetMat`. Note that the mapping is *not* overwritten here. In the second step, we map `argMat`, not `origArg`.

And we also generate two `unrealized_conversion_cast` ops. Such casts cannot be folded, unless they `type(origArg) == type(targetMat)`.

It was actually on purpose that two casts in a row are generated here. The first one is an argument materialization and the second one is target materialization. Depending on the configuration of the type converter, we may not generate `unrealized_conversion_cast`, but custom ops.